core/include/detray/geometry/surfaces_finder.hpp

```diff
@@ -0,0 +1,137 @@
+          template <typename...> class tuple_type = dtuple,
+          template <typename...> class vector_type = dvector,
+          template <typename...> class jagged_vector_type = djagged_vector>
+struct surfaces_finder {
+    public:
+    surfaces_finder() = delete;
+
+    using surfaces_regular_circular_grid =
+        grid2<attach_populator, axis::regular, axis::circular, serializer2,
+              vector_type, jagged_vector_type, array_type, tuple_type, dindex,
+              false>;
+
+    static const size_t N_GRIDS = N;
+
+    /// host constructor
+    template <std::size_t... ints>
+    DETRAY_HOST surfaces_finder(vecmem::memory_resource& resource)
+        : _surface_grids(host(resource, std::make_index_sequence<N_GRIDS>{})) {}
+
+    template <std::size_t... ints>
+    DETRAY_HOST auto host(vecmem::memory_resource& resource,
+                          std::index_sequence<ints...> /*seq*/) {
+
+        // TODO: Are raw pointers OK here?
+        array_type<vecmem::memory_resource*, N_GRIDS> resources;
+
+        std::fill(resources.begin(), resources.end(), &resource);
+        return darray<surfaces_regular_circular_grid, N_GRIDS>(
+            {{*resources[ints]...}});
+    }
+
+    /// device constructor
+    template <typename surfaces_finder_data_type,
+              std::enable_if_t<!std::is_base_of_v<vecmem::memory_resource,
+                                                  surfaces_finder_data_type>,
+                               bool> = true>
+    DETRAY_HOST_DEVICE surfaces_finder(surfaces_finder_data_type& finder_data)
+        : _surface_grids(
+              device(finder_data, std::make_index_sequence<N_GRIDS>{})) {}
+
+    template <std::size_t... ints, typename surfaces_finder_data_type>
+    DETRAY_HOST_DEVICE auto device(surfaces_finder_data_type& finder_data,
+                                   std::index_sequence<ints...> /*seq*/) {
+        return darray<surfaces_regular_circular_grid, N_GRIDS>(
+            {{finder_data[ints]...}});
+    }
+
+    DETRAY_HOST_DEVICE size_t size() { return N_GRIDS; }
+
+    /** Access operator - non-const
+     * @return the surface grid element
+     */
+    DETRAY_HOST_DEVICE
+    auto& operator[](unsigned int value_index) {
+        return _surface_grids[value_index];
+    }
+
+    darray<surfaces_regular_circular_grid, N_GRIDS> _surface_grids;
+};
+
+/** A static implemetation of surfaces_finder data for device
+ */
+template <typename surfaces_finder_type>
+struct surfaces_finder_data {
+    // type definition
+    using surfaces_finder_t = surfaces_finder_type;
+    using surface_grid_t =
+        typename surfaces_finder_t::surfaces_regular_circular_grid;
+
+    surfaces_finder_data(surfaces_finder_type& finder,
+                         vecmem::memory_resource& resource)
+        : _surface_grids_data(
+              make_grid_data_array(finder._surface_grids, resource)) {}
+
+    darray<grid2_data<surface_grid_t>, surfaces_finder_type::N_GRIDS>
```

Comment:
Now... If `detray::surfaces_finder` properly declares its "array type", you will be able to use that here as well. Instead of hardcoding `darray`.

Reply:
right. fixed it now